### PR TITLE
Improve window_funnel through Fix bug, use const_column and add DATE type

### PR DESCRIPTION
**Pull request by Pslydhh**
## What type of PR is this：
- [x] bug
- [x] feature
- [ ] enhancement
- [ ] others

Based on https://github.com/StarRocks/starrocks/pull/4985
## Problem Summary(Required) ：

- Fixes bug caused by un-sorted event_list.
- Use const_column for window_size to avoid serialize/deserialize.
- Add DATE TYPE as time parameter.

## use case(datetime) : 
```
mysql> select * from action;
+------+------------+---------------------+
| uid  | event_type | time                |
+------+------------+---------------------+
| 1    | 浏览       | 2020-01-02 11:00:00 |
| 1    | 点击       | 2020-01-02 11:10:00 |
| 1    | 下单       | 2020-01-02 11:20:00 |
| 1    | 支付       | 2020-01-02 11:30:00 |
| 1    | 浏览       | 2020-01-02 11:00:00 |
| 2    | 下单       | 2020-01-02 11:00:00 |
| 2    | 支付       | 2020-01-02 11:10:00 |
| 3    | 浏览       | 2020-01-02 11:20:00 |
| 3    | 点击       | 2020-01-02 12:00:00 |
| 4    | 浏览       | 2020-01-02 11:50:00 |
| 4    | 点击       | 2020-01-02 12:00:00 |
| 5    | 浏览       | 2020-01-02 11:50:00 |
| 5    | 点击       | 2020-01-02 12:00:00 |
| 5    | 下单       | 2020-01-02 11:10:00 |
| 6    | 浏览       | 2020-01-02 11:50:00 |
| 6    | 点击       | 2020-01-02 12:00:00 |
| 6    | 下单       | 2020-01-02 12:10:00 |
+------+------------+---------------------+
17 rows in set (0.01 sec)

mysql> select uid, window_funnel(600, time, 0, [event_type="浏览", event_type="点击", event_type="下单", event_type="支付"]) as level from action group by uid order by uid;
+------+-------+
| uid  | level |
+------+-------+
| 1    |     2 |
| 2    |     0 |
| 3    |     1 |
| 4    |     2 |
| 5    |     2 |
| 6    |     2 |
+------+-------+
6 rows in set (0.04 sec)
```

## use case(date) : 
```
mysql> select * from action9;
+------+------------+------------+
| uid  | event_type | time       |
+------+------------+------------+
| 1    | 浏览       | 2020-01-02 |
| 1    | 点击       | 2020-01-03 |
| 1    | 下单       | 2020-01-04 |
| 1    | 支付       | 2020-01-05 |
| 2    | 浏览       | 2020-01-02 |
| 2    | 点击       | 2020-01-06 |
+------+------------+------------+
6 rows in set (0.01 sec)

mysql> select uid, window_funnel(3, time, 0, [event_type="浏览", event_type="点击", event_type="下单", event_type="支付"]) as level from action9 group by uid order by uid;
+------+-------+
| uid  | level |
+------+-------+
| 1    |     4 |
| 2    |     1 |
+------+-------+
2 rows in set (0.02 sec)
```

### CODE REVIEW

File: be/src/exprs/agg/window_funnel.h [hunk 1]
```diff
@@ -135,16 +143,25 @@ struct WindowFunnelState {
...
 public:
     void update(FunctionContext* ctx, const Column** columns, AggDataPtr __restrict state, size_t row_num) const {
-        const auto* window_size_column = down_cast<const Int64Column*>(columns[0]);
-        this->data(state).window_size = window_size_column->get(row_num).get_int64();
+        this->data(state).window_size = ColumnHelper::get_const_value<TYPE_BIGINT>(ctx->get_constant_column(0));
 
         // get timestamp
-        const auto* timestamp_column = down_cast<const TimestampColumn*>(columns[1]);
-        auto tv = timestamp_column->get(row_num).get_timestamp();
+        const auto* timestamp_column = down_cast<const TimeTypeColumn*>(columns[1]);
```
Comment (stdpain) [inline be/src/exprs/agg/window_funnel.h:157]: columns[1] may be a const column. because we only unpack the first parameter.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|

**Comment by stdpain** [comment]
@Pslydhh could you add an example for how to call window_funnel?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|1|1|100.00%|

**Comment by Pslydhh** [comment]
> @Pslydhh could you add an example for how to call window_funnel?

I have add use case for datetime and date.

**Review by stdpain — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**